src/serverenvironment.cpp

```diff
@@ -254,40 +254,72 @@ void LBMManager::applyLBMs(ServerEnvironment *env, MapBlock *block,
 	FATAL_ERROR_IF(!m_query_mode,
 		"attempted to query on non fully set up LBMManager");
 
-	const v3s16 pos_of_block = block->getPosRelative();
-	v3s16 pos;
-	MapNode n;
-	content_t c;
-	auto it = getLBMsIntroducedAfter(stamp);
+	// Collect a list of all LBMs and associated positions
+	struct LBMToRun {
+		std::unordered_set<v3s16> p; // node positions
+		std::unordered_set<LoadingBlockModifierDef*> l;
+	};
+	std::unordered_map<content_t, LBMToRun> to_run;
+
 	// Note: the iteration count of this outer loop is typically very low, so it's ok.
-	for (; it != m_lbm_lookup.end(); ++it) {
+	for (auto it = getLBMsIntroducedAfter(stamp); it != m_lbm_lookup.end(); ++it) {
+		v3s16 pos;
+		content_t c;
+
 		// Cache previous lookup result since it has a high performance penalty.
 		content_t previous_c = CONTENT_IGNORE;
 		const LBMContentMapping::lbm_vector *lbm_list = nullptr;
 
 		for (pos.Z = 0; pos.Z < MAP_BLOCKSIZE; pos.Z++)
 		for (pos.Y = 0; pos.Y < MAP_BLOCKSIZE; pos.Y++)
 		for (pos.X = 0; pos.X < MAP_BLOCKSIZE; pos.X++) {
-			n = block->getNodeNoCheck(pos);
-			c = n.getContent();
+			c = block->getNodeNoCheck(pos).getContent();
 
+			bool c_changed = false;
 			if (previous_c != c) {
+				c_changed = true;
 				lbm_list = it->second.lookup(c);
 				previous_c = c;
 			}
 
 			if (!lbm_list)
 				continue;
-			for (auto lbmdef : *lbm_list) {
-				lbmdef->trigger(env, pos + pos_of_block, n, dtime_s);
-				if (block->isOrphan())
-					return;
-				n = block->getNodeNoCheck(pos);
-				if (n.getContent() != c)
-					break; // The node was changed and the LBMs no longer apply
+			auto &batch = to_run[c];
+			batch.p.insert(pos);
+			if (c_changed) {
+				batch.l.insert(lbm_list->begin(), lbm_list->end());
```

Comment:
Insert only needed for new batches (i.e. when `to_run.insert()` returns [it,true]).

Reply:
this assumption does not hold due to the outer loop